In [4]:
import numpy as np

n = np.array([1,2,3,4])

In [6]:
#shape: uma tupla que contém o número de elementos (tamanho) em cada dimensão do array;
n.shape

(4,)

In [7]:
#size: número total de elementos do array;
n.size

4

In [9]:
#ndim: número de dimensões (eixos);
n.ndim

1

In [10]:
#nbytes: número de bytes utilizado no armazenamento dos dados
n.nbytes

32

In [11]:
dtype: data type dos elementos do array.
n.dtype

dtype('int64')

In [15]:
print(type(n))

<class 'numpy.ndarray'>


In [17]:
n.info

AttributeError: 'numpy.ndarray' object has no attribute 'info'

# Exercício 2

Com base no exercício anterior crie uma função que receba um array e mostre na tela seus atributos. Verifique se o argumento inserido é de fato um objeto numpy. Use também a função built-in type() para verificar a classe do objeto.

Retorno esperado:

inspecionar_array(np.array([1,2,3]))
 
- Shape : (3,)
- Size:  3
- Ndim:  1
- Nbytes:  24
- Dtype:  int64
- Type:  <class 'numpy.ndarray'>
- array([1, 2, 3, 4])


- inspecionar_array([1,2,3])
 
'O objeto inserido não é um array NumPy.'

In [5]:
def inspecionar_array(array):
    if isinstance(array, np.ndarray):
        print('Shape: ', array.shape)
        print('Size: ', array.size)
        print('Ndim: ', array.ndim)
        print('Nbytes: ', array.nbytes)
        print('Dtype: ', array.dtype)
        print('Type: ', type(array))
        return array
    else:
        return 'O objeto inserido não é um array NumPy.'

In [2]:
# Criando um array unimidensional
array = np.array([1,2,3,4])

In [3]:
# Criando uma simples lista
list_ = [1,2,3,4]

In [6]:
# chamando a função inspecinar_array para conferir se o objeto é um array
inspecionar_array(array)

Shape:  (4,)
Size:  4
Ndim:  1
Nbytes:  32
Dtype:  int64
Type:  <class 'numpy.ndarray'>


array([1, 2, 3, 4])

In [7]:
inspecionar_array(list_)

'O objeto inserido não é um array NumPy.'

# Exercício 3

Use a função np.array() para criar os arrays unidimensionais listados abaixo:


a) Use uma lista como argumento e obtenha:

In [10]:
import numpy as np

lista = [1,2,3,4]

arr = np.array([lista])

arr

array([[1, 2, 3, 4]])

Use a função range() e obtenha:

In [12]:
import numpy as np

arr = np.array(range(1,11))
arr

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

c)Use uma tupla como argumento e obtenha:

In [16]:
tupla = (10,11,12,13,14)
array3 = np.array(tupla)
array3

array([10, 11, 12, 13, 14])

d) Use a função range() e obtenha:

array([2020, 2010, 2000, 1990, 1980, 1970, 1960, 1950, 1940, 1930, 1920,
           1910, 1900])

In [13]:
# Exibe apenas o resultado da operação porém não foi armazenado em uma variavel
np.array(range(2020,1899,-10 ))

array([2020, 2010, 2000, 1990, 1980, 1970, 1960, 1950, 1940, 1930, 1920,
       1910, 1900])

e) Use a função range() e obtenha:

array([1990, 1995, 2000, 2005, 2010, 2015, 2020])

In [14]:
np.array(range(1990, 2021, 5))

array([1990, 1995, 2000, 2005, 2010, 2015, 2020])

f) Obtenha:
array(100)

In [15]:
np.array(100)

array(100)

# Exercício 4

A função np.ones() permite criar um array preenchido com o número 1 em todas as suas entradas. Podemos especificar como argumentos o shape e o tipo de dado dtype do array. Use esta função para criar os arrays a seguir:

a)

- array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [22]:
import numpy as np

one = np.ones(10)
one

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

## b)



array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [26]:
# Não esquecer de colocar em dois parênteses
one1 = np.ones((4,4))
one1

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

### c)
 
C = np.ones((3,2), dtype=float)

In [31]:
# Números 1 convertidos em float
np.ones((4,2), dtype = float)

array([[1., 1.],
       [1., 1.],
       [1., 1.],
       [1., 1.]])

# Exercício 5
A função np.zeros() permite criar um array preenchido com o número 0 em todas as suas entradas. Podemos especificar como argumentos o shape e o tipo de dado dtype do array (default é o float). Use esta função para criar os arrays a seguir:

### a)



array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [33]:
import numpy as np

zero = np.zeros(10)
zero

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

### b)



array([[0., 0.],
       [0., 0.]]) 

In [37]:
np.zeros((2,2), dtype = int)

array([[0, 0],
       [0, 0]])

In [39]:
np.zeros((4,6), dtype = float)

array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]])

# Exercício 6

Crie os arrays a seguir:

   

### a)

array([[10],
       [15],
       [20],
       [25],
       [30],
       [35],
       [40],
       [45]])

In [45]:
import numpy as np

np.arange(10,50,5).reshape(-1,1)

array([[10],
       [15],
       [20],
       [25],
       [30],
       [35],
       [40],
       [45]])

### b)



array([-50, -45, -40, -35, -30, -25, -20, -15, -10,  -5])

In [47]:
np.arange(-50, 0, 5)

array([-50, -45, -40, -35, -30, -25, -20, -15, -10,  -5])

In [48]:
#Assim da erro temos que usar dois parênteses
np.arange(50,150,2).reshape(5,10)

ValueError: cannot reshape array of size 49 into shape (5,10)

In [50]:
np.arange(50,150,2).reshape((5,10))

array([[ 50,  52,  54,  56,  58,  60,  62,  64,  66,  68],
       [ 70,  72,  74,  76,  78,  80,  82,  84,  86,  88],
       [ 90,  92,  94,  96,  98, 100, 102, 104, 106, 108],
       [110, 112, 114, 116, 118, 120, 122, 124, 126, 128],
       [130, 132, 134, 136, 138, 140, 142, 144, 146, 148]])

# Exercício 7

Utilize a função np.full(shape, fill_value, dtype) para criar os arrays a seguir. Esta função permite criar um array com um shape e dtype definidos pelo usuário, preenchido com um determinado valor.

### a)



array([5., 5., 5., 5., 5., 5., 5., 5., 5., 5.])

In [55]:
import numpy as np

np.full(shape = 10, fill_value = 5, dtype = int)

array([5, 5, 5, 5, 5, 5, 5, 5, 5, 5])

### b)



array([[-5., -5., -5., -5.],
       [-5., -5., -5., -5.],
       [-5., -5., -5., -5.],
       [-5., -5., -5., -5.]])

In [56]:
np.full(shape = (3,3), fill_value = 3, dtype = float)

array([[3., 3., 3.],
       [3., 3., 3.],
       [3., 3., 3.]])

#### d

array([[50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50]])

In [57]:
np.full((10,10), 50, int)

array([[50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50, 50, 50, 50, 50]])

Dica: outra possibilidade seria obter os arrays com uma operação diferente, a partir da multiplicação de uma matriz preenchida com número 1 por um determinado valor.

# Exercício 8
Defina o array K e verifique se há dados ausentes (missing data). Use a função np.isnan().



array([[ nan,   1.,  10.,   0.],
       [ nan,  nan,  nan,  nan],
       [100.,  50.,  nan, -25.],
       [ 30.,  nan,  nan, 130.]])

In [67]:
import numpy as np

k = np.array([[ np.nan,   1.,  10.,   0.],
       [ np.nan,  np.nan,  np.nan,  np.nan],
       [100.,  50.,  np.nan, -25.],
       [ 30.,  np.nan,  np.nan, 130.]
])

In [68]:
k

array([[ nan,   1.,  10.,   0.],
       [ nan,  nan,  nan,  nan],
       [100.,  50.,  nan, -25.],
       [ 30.,  nan,  nan, 130.]])

In [69]:
np.isnan(k)

array([[ True, False, False, False],
       [ True,  True,  True,  True],
       [False, False,  True, False],
       [False,  True,  True, False]])

In [65]:
k = np.isnan(k)

In [66]:
k

array([[ True, False, False, False],
       [ True,  True,  True,  True],
       [False, False,  True, False],
       [False,  True,  True, False]])

In [71]:
print(np.isnan(k))

[[ True False False False]
 [ True  True  True  True]
 [False False  True False]
 [False  True  True False]]


In [73]:
# Soma a quantidade de dados ausentes em k
print(f'Há {np.isnan(k).sum()} dados ausentes em k')

Há 8 dados ausentes em k


# Exercício 9
A função np.linspace() retorna um array com números contendo espaçamento uniforme em um intervalo definido. Crie um array com números dentro do intervalo [1,21] em 5 partes.

Retorno esperado:



array([ 1.,  6., 11., 16., 21.])

In [75]:
import numpy as np

np.linspace(1,21,5)

array([ 1.,  6., 11., 16., 21.])

# Exercício 10
A função np.logspace() retorna números com espaçamento uniforme em uma escala logarítmica. Crie um array com números dentro do intervalo [1, 51] em 10 partes com a função np.logspace().

Retorno esperado:



array([[1.00000000e+01],
       [3.59381366e+06],
       [1.29154967e+12],
       [4.64158883e+17],
       [1.66810054e+23],
       [5.99484250e+28],
       [2.15443469e+34],
       [7.74263683e+39],
       [2.78255940e+45],
       [1.00000000e+51]])

In [77]:
import numpy as np

np.logspace(1,51,10).reshape(-1,1)

array([[1.00000000e+01],
       [3.59381366e+06],
       [1.29154967e+12],
       [4.64158883e+17],
       [1.66810054e+23],
       [5.99484250e+28],
       [2.15443469e+34],
       [7.74263683e+39],
       [2.78255940e+45],
       [1.00000000e+51]])

# Exercício 11
Use a função np.diag() para construir o array a seguir:



array([[ 5,  0,  0,  0,  0,  0],
       [ 0, 10,  0,  0,  0,  0],
       [ 0,  0, 15,  0,  0,  0],
       [ 0,  0,  0, 20,  0,  0],
       [ 0,  0,  0,  0, 25,  0],
       [ 0,  0,  0,  0,  0, 30]])

In [78]:
import numpy as np

np.diag(range(5,31,5))

array([[ 5,  0,  0,  0,  0,  0],
       [ 0, 10,  0,  0,  0,  0],
       [ 0,  0, 15,  0,  0,  0],
       [ 0,  0,  0, 20,  0,  0],
       [ 0,  0,  0,  0, 25,  0],
       [ 0,  0,  0,  0,  0, 30]])

# Exercício 12
Use a função np.eye() (ou ainda np.identity()) para obter o array seguir.



array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]])


Dica: essa matriz é definida como matriz identidade, neste caso de ordem 10, e assume grande relevância em diversos cálculos computacionais, como veremos mais adiante.

In [79]:
import numpy as np
np.eye(10)

array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]])

In [80]:
np.identity(n = 10)

array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]])

# Exercício 14
Obtenha a soma dos elementos de cada array a seguir. Use a função np.sum(). Implemente uma solução com loop for, para tanto armazene os vetores em uma lista.



import numpy as np
array1 = np.arange(1,11)
array2 = np.linspace(1,50,10)
array3 = np.array(range(51))
array4 = np.logspace(1,10,5)
array5 = np.diag(range(5,31,5))


Retorno esperado:



Soma dos elementos do array 1:  55
Soma dos elementos do array 2:  255.0
Soma dos elementos do array 3:  1275
Soma dos elementos do array 4:  10056552148.564463

In [ ]:
import numpy as np



In [83]:
array1 = np.arange(1,11)
array1

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [85]:
array2 = np.linspace(1,50,10)
array2

array([ 1.        ,  6.44444444, 11.88888889, 17.33333333, 22.77777778,
       28.22222222, 33.66666667, 39.11111111, 44.55555556, 50.        ])

In [87]:
array3 = np.array(range(51))
array3

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])

In [88]:
array4 = np.logspace(1,10,15)
array4

array([1.00000000e+01, 4.39397056e+01, 1.93069773e+02, 8.48342898e+02,
       3.72759372e+03, 1.63789371e+04, 7.19685673e+04, 3.16227766e+05,
       1.38949549e+06, 6.10540230e+06, 2.68269580e+07, 1.17876863e+08,
       5.17947468e+08, 2.27584593e+09, 1.00000000e+10])

In [89]:
array5 = np.diag(range(5,31,5))


In [99]:
arrays = [array1,array2,array3,array4,array5]
arrays

[array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10]),
 array([ 1.        ,  6.44444444, 11.88888889, 17.33333333, 22.77777778,
        28.22222222, 33.66666667, 39.11111111, 44.55555556, 50.        ]),
 array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
        34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]),
 array([1.00000000e+01, 4.39397056e+01, 1.93069773e+02, 8.48342898e+02,
        3.72759372e+03, 1.63789371e+04, 7.19685673e+04, 3.16227766e+05,
        1.38949549e+06, 6.10540230e+06, 2.68269580e+07, 1.17876863e+08,
        5.17947468e+08, 2.27584593e+09, 1.00000000e+10]),
 array([[ 5,  0,  0,  0,  0,  0],
        [ 0, 10,  0,  0,  0,  0],
        [ 0,  0, 15,  0,  0,  0],
        [ 0,  0,  0, 20,  0,  0],
        [ 0,  0,  0,  0, 25,  0],
        [ 0,  0,  0,  0,  0, 30]])]

In [98]:

for index, array in enumerate(arrays, start = 1):
    print(f'Soma dos elementos do array {index}', array.sum())

Soma dos elementos do array 1 55
Soma dos elementos do array 2 255.0
Soma dos elementos do array 3 1275
Soma dos elementos do array 4 12946401511.437511
Soma dos elementos do array 5 105


# Exercício 15
A partir da matriz definida abaixo obtenha: a) soma total dos elementos; b) soma das linhas ; c) soma das colunas.



import numpy as np

matriz = np.array(np.random.randint(1,100,15)).reshape(5,3)



In [100]:
import numpy as np

matriz = np.array(np.random.randint(1,100,15).reshape(5,3))
matriz

array([[61, 57, 32],
       [11, 25, 72],
       [34, 16, 21],
       [66,  3, 52],
       [74, 82, 34]])

In [105]:
print('Soma total dos elementos: ', matriz.sum())

Soma total dos elementos:  640


In [107]:
print('Soma total das linhas: ', matriz.sum(axis = 1))

Soma total das linhas:  [150 108  71 121 190]


In [108]:
print('Soma total das colunas: ', matriz.sum(axis = 0))

Soma total das colunas:  [246 183 211]


# Exercício 16
Obtenha o valor máximo e o valor mínimo de cada array abaixo. Implemente uma solução com loop for.



- import numpy as np
- array1 = np.array([-1, 10, 3, 4, 7, 27])
- array2 = np.random.randint(-100,300,30)
-  array3 = np.random.randn(10)
- array4 = np.array(range(50,10,-2))


Dica: para obter o valor máximo e mínimo de um array podemos usar array.max() e array.min() ( ou ainda max(array) e min(array)).

In [109]:
import numpy as np

array1 = np.array([-1,10,3,4,7,27])
array2 = np.random.randint(-100,300,30)
array3 = np.random.randn(10)
array4 = np.array(range(50,10,-2))
arrays = [array1,array2, array3,array4]

In [114]:
for index, elemento in enumerate(arrays):
    print(f'Array {index}')
    print(f'Valor máximo {elemento.max()}')
    print(f'Valor mínimo {elemento.min()}')
    print()

Array 0
Valor máximo 10
Valor mínimo 1

Array 1
Valor máximo 50.0
Valor mínimo 1.0

Array 2
Valor máximo 50
Valor mínimo 0

Array 3
Valor máximo 10000000000.0
Valor mínimo 10.0

Array 4
Valor máximo 30
Valor mínimo 0



# Exercício 17
Crie uma função que receba uma matriz como entrada e calcule:

-    a) Valor máximo da matriz;

-    b) Valor máximo por coluna;

-    c) Valor máximo por linha;

-    d) Valor mínimo da matriz;

 -   e) Valor mínimo por coluna;

 -   f) Valor mínimo por linha.



Exemplo de retorno para a matriz definida abaixo:

   

Input:

matriz = np.random.randint(-1000,1000,50).reshape(10,5)
info_matriz(matriz)
Output:

- Valor máximo da matriz:  987
- Valor máximo da matriz:  [959 987 597 748 719]
- Valor máximo por linha:  [ 987  312  982  204  719  689  101  793 -460  959]
- Valor máximo por linha:  -969
- Valor mínimo por coluna:  [-878 -966 -969 -884 -881]
- Valor mínimo por linha:  [  30 -878 -335 -863  -40 -289 -660 -239 -969 -966]

In [119]:
def func(matrix):
    if isinstance(matriz, np.ndarray) and matrix.ndim == 2:
        print(f'Valor máximo da matriz: {matrix.max()}')
        print(f'Valor máximo da matriz: {matrix.max(axis = 0)}')
        print(f'Valor máximo por linha: {matrix.max(axis = 1)}')
        print(f'Valor máximo por linha: {matrix.min()}')
        print(f'Valor mínimo por coluna: {matrix.min(axis = 0)}')
        print(f'Valor mínimo por linha: {matrix.min(axis = 1)}')
    else:
        return None

In [117]:
matriz = np.random.randint(-1000,1000,50).reshape(10,5)
matriz

array([[ 780,  -80, -943, -742, -996],
       [ -33, -773,  -33, -817, -271],
       [ 398, -646,  125,  857,  779],
       [ 223,  824,  556, -721, -632],
       [-864,  806,  -49, -814, -631],
       [ 945, -789,  -60,  506, -882],
       [ 774, -693,  -64,   17,  865],
       [-246,  556, -395,  473,   30],
       [ 336,  459, -379,  247,   84],
       [ 180, -457,  182,  559, -817]])

In [124]:
if __name__ == '__main__':
    func(matriz)

Valor máximo da matriz: 945
Valor máximo da matriz: [945 824 556 857 865]
Valor máximo por linha: [780 -33 857 824 806 945 865 556 459 559]
Valor máximo por linha: -996
Valor mínimo por coluna: [-864 -789 -943 -817 -996]
Valor mínimo por linha: [-996 -817 -646 -721 -864 -882 -693 -395 -379 -817]


In [125]:
lista = [1,2,3,5]

In [126]:
func(lista)

AttributeError: 'list' object has no attribute 'max'

# Exercício 18
Com base no array definido abaixo obtenha sua média, mediana, variância e desvio padrão.



import numpy as np
arr = np.random.normal(40,0.5,size = (300))


Dica:

- Média: np.mean();

- Mediana: np.median();

- Desvio padrão: np.std();

- Variância: np.var().

In [128]:
import numpy as np

arr = np.random.normal(40, 0.5, size = (300))
arr

array([39.65954583, 40.71785386, 38.77645972, 40.93172914, 39.80495428,
       40.21619726, 38.63049418, 41.00585094, 39.28413766, 39.60122932,
       40.09214686, 39.4738071 , 40.08446699, 40.3157963 , 40.23451529,
       39.6349439 , 40.38242454, 39.3227198 , 40.04719448, 39.99256774,
       39.27836343, 39.75927829, 40.16027774, 39.36180497, 39.68707558,
       39.99426514, 40.14519733, 39.92111327, 39.97707806, 39.30558998,
       40.1987879 , 40.1211927 , 39.8758513 , 40.63222854, 40.27985475,
       40.96340934, 40.43567639, 40.5467907 , 39.6068207 , 40.13534083,
       40.73923609, 39.01290987, 40.66334308, 40.62153938, 39.44652008,
       39.68329584, 39.57339508, 39.99176425, 40.41452005, 40.25343958,
       40.60721319, 40.51889938, 40.24566801, 40.1665863 , 39.48758876,
       40.17332748, 40.18066177, 39.99997365, 39.44287133, 40.14996327,
       39.56835888, 40.00077544, 40.9288927 , 40.35624141, 40.3950229 ,
       40.22645074, 40.67120751, 39.89438788, 39.49183683, 40.96

In [129]:
arr.mean()

40.02308244524398

In [135]:
np.median(arr)

40.02336519360387

In [133]:
arr.std()

0.5185025678858931

In [134]:
arr.var()

0.2688449129042652

In [162]:
arr.median

AttributeError: 'numpy.ndarray' object has no attribute 'median'

# Exercício 19
Com base nos arrays definidos a seguir obtenha a matriz de covariância e a matriz de correlação.



import numpy as np
x = np.random.rand(100)
y = np.random.uniform(size=100)


Dica: use np.corrcoef() e np.cov().

In [137]:
import numpy as np

x = np.random.rand(100)
y = np.random.uniform(size = 100)

In [138]:
x

array([0.31137385, 0.95466808, 0.45151103, 0.78067212, 0.82282532,
       0.67755819, 0.88167246, 0.31829911, 0.94616612, 0.72679685,
       0.99004402, 0.82983784, 0.52436298, 0.68215138, 0.39175858,
       0.32004858, 0.37809923, 0.64910477, 0.35744567, 0.95351657,
       0.65744011, 0.39258256, 0.01115301, 0.70672697, 0.69944048,
       0.77833798, 0.648949  , 0.82177529, 0.34268485, 0.41773072,
       0.83794388, 0.31238663, 0.49546807, 0.83369744, 0.91517691,
       0.22698978, 0.56345186, 0.16724734, 0.85140407, 0.23198644,
       0.09991333, 0.67456127, 0.91007226, 0.26001043, 0.01170932,
       0.27425383, 0.6977672 , 0.79556468, 0.61094013, 0.48670024,
       0.99364133, 0.3306753 , 0.60960463, 0.22174897, 0.69716544,
       0.33935237, 0.78125411, 0.451125  , 0.93229112, 0.34492231,
       0.51586331, 0.89830183, 0.38438047, 0.25786485, 0.80644818,
       0.67465599, 0.35017295, 0.81286674, 0.49138501, 0.34611037,
       0.83123882, 0.34003156, 0.85791385, 0.01717917, 0.94025

In [139]:
y

array([3.31634043e-01, 2.26565475e-01, 8.26874512e-01, 3.34465268e-01,
       2.42737867e-01, 6.58377378e-02, 8.67787690e-01, 3.30488561e-01,
       5.95120537e-01, 1.16828963e-01, 8.40060768e-01, 8.19218928e-01,
       6.89240610e-01, 2.37973838e-01, 1.81149788e-01, 7.04788652e-01,
       6.61976157e-02, 9.62755654e-01, 8.58950944e-01, 5.31494322e-01,
       6.10727377e-01, 8.05744956e-01, 6.82887860e-01, 3.31895061e-02,
       1.58050426e-01, 8.94215158e-01, 8.24260062e-01, 6.05897052e-01,
       7.57599262e-01, 1.03068271e-01, 7.91539589e-01, 5.63201791e-01,
       3.60209519e-01, 5.87297440e-01, 2.96288066e-02, 6.70442268e-01,
       2.14227571e-01, 4.23955621e-04, 9.40773646e-01, 1.32550353e-01,
       2.19525635e-01, 7.78026036e-01, 5.42180714e-01, 3.95959314e-01,
       1.03607583e-01, 8.74150973e-01, 7.16672034e-01, 2.19741409e-01,
       8.56170534e-02, 4.42373106e-01, 2.33665466e-01, 1.10361382e-01,
       3.39488839e-01, 5.22290816e-02, 4.88209914e-01, 7.84229233e-01,
      

In [142]:
np.corrcoef(x, y)

array([[1.        , 0.12832299],
       [0.12832299, 1.        ]])

In [145]:
np.cov(x, y)

array([[0.07856164, 0.01114565],
       [0.01114565, 0.09602649]])

# Exercício 20
Obtenha o índice do valor máximo do array A.



A = np.array([[11, 15, 33, 105],
              [1, 140, 45, 90],
              [67, 230, 78, 99]])


Dica: use np.argmax() ou array.argmax().

In [146]:
import numpy as np

a = np.array(
    [[11, 15, 33, 105],
    [1, 140, 45, 90],
    [67, 230, 78, 99]]
)
a

array([[ 11,  15,  33, 105],
       [  1, 140,  45,  90],
       [ 67, 230,  78,  99]])

In [150]:
print(f' O valor maximo de a é: {a.argmax()}')

 O valor maximo de a é: 9


In [148]:
np.argmax(a)

9

# Obtenha o índice do valor mínimo do array B.



B = np.array([
    [73, 22, 20, 62,  3],
    [68, 84, 27, 59, 93],
    [48, 52,  4, 55, 50],
    [42, 23, 63, 67,  4],
    [44, 30, 41, 31, 29]])


Dica: use np.argmin() ou array.argmin().

In [151]:
B = np.array(
    [[73, 22, 20, 62,  3],
    [68, 84, 27, 59, 93],
    [48, 52,  4, 55, 50],
    [42, 23, 63, 67,  4],
    [44, 30, 41, 31, 29]]
)

B

array([[73, 22, 20, 62,  3],
       [68, 84, 27, 59, 93],
       [48, 52,  4, 55, 50],
       [42, 23, 63, 67,  4],
       [44, 30, 41, 31, 29]])

In [152]:
print(f'O válor do índice no menor valor de B é: {B.argmin()}')

O válor do índice no menor valor de B é: 4


In [154]:
print(f'O válor do índice no menor valor de B é: {np.argmin(B)}')

O válor do índice no menor valor de B é: 4


# Exercício 22
Escreva uma função que receba um array NumPy e mostre na tela um conjunto de  estatísticas descritivas deste objeto.



Resultado esperado:

Input

- array = np.array([-57, 101, 270, 130, 144])
- info_array(array)


Output:

- Valor máximo: 270
- Valor mínimo: -57
- Média: 117.6
- Mediana: 130.0
- Variância: 10967.439999999999
- Desvio padrão: 104.72554607162475

In [168]:
import numpy as np

def mostra_estatisticas(array):
    if isinstance(array, np.ndarray):
        print(f'Valor Máximo: {array.max()}')
        print(f'Valor Minímo: {array.min()}')
        print(f'Média: {array.mean()}')
        print(f'Mediana: {np.median(array)}')
        print(f'Variancia: {array.var()}')
        print(f'Desvio Padrão: {array.std}')
        return type(array)
    else:
        return None

In [167]:
B = np.array(
    [[73, 22, 20, 62,  3],
    [68, 84, 27, 59, 93],
    [48, 52,  4, 55, 50],
    [42, 23, 63, 67,  4],
    [44, 30, 41, 31, 29]]
)

In [156]:
type(B)

numpy.ndarray

In [170]:
mostra_estatisticas(B)

Valor Máximo: 93
Valor Minímo: 3
Média: 43.76
Mediana: 44.0
Variancia: 574.6623999999999
Desvio Padrão: <built-in method std of numpy.ndarray object at 0x7f59129282d0>


numpy.ndarray

# Exercício 23
Use a função np.unique() para obter um array com os elementos únicos do array abaixo. Para tanto, armazene o novo array em uma variável.



A = np.array([
    [1, 3, 3, 4, 5],
    [4, 7, 11, 3, 5],
    [1, 11, 4, 4, 11]])

In [171]:
import numpy as np

A = np.array([
    [1,3,3,4,5],
    [4,7,11,3,5],
    [1,11,4,4,11]
])

A

array([[ 1,  3,  3,  4,  5],
       [ 4,  7, 11,  3,  5],
       [ 1, 11,  4,  4, 11]])

In [173]:
np.unique(A)

array([ 1,  3,  4,  5,  7, 11])

# Exercício 24
Indexação de arrays 1-D



Considerando os arrays abaixo:



- import numpy as np
- array1 = np.array([10, 20, 30, 40])
- array2 = np.array([100, 30, 5, 1, 40, 100, 20, 130, 155,170])
   

Obtenha:

a) o primeiro, o terceiro e o último elemento de cada array;

b) inverta o array;

c) pegue o número 20 em ambos os arrays e armazene a soma em uma variável.

In [174]:
import numpy as np

array1 = np.array([10, 20, 30, 40])
array2 = np.array([100, 30, 5, 1, 40, 100, 20, 130, 155,170])

In [178]:
# Obtenha: o primeiro, o terceiro e o último elemento de cada array
lista = [array1, array2]

for index, i in enumerate(lista, start =1):
    print(f'Array {index}')
    print(f'Primeiro elemento {i[0]}')
    print(f'Terceiro elemento: {i[2]}')
    print(f'último elemento: {i[-1]}')
    print(f'array invertido: {i[::-1]}')



Array 1
Primeiro elemento 10
Terceiro elemento: 30
último elemento: 40
array invertido: [40 30 20 10]
Array 2
Primeiro elemento 100
Terceiro elemento: 5
último elemento: 170
array invertido: [170 155 130  20 100  40   1   5  30 100]


In [180]:
soma = array1[1] + array2[6]
soma


40

# Exercício 25
Indexação de arrays 2-D



Acesse os números 9 e 90 do array abaixo.



- import numpy as np
- ar = np.array([[-1, 33, 44, 9, 1],[11, 1, 0, 90, 44]])

In [183]:
import numpy as np

ar = np.array([[-1, 33, 44, 9, 1],[11, 1, 0, 90, 44]])
ar

array([[-1, 33, 44,  9,  1],
       [11,  1,  0, 90, 44]])

In [190]:
print(ar[0, -2])
print(ar[1, -2])

9
90


# Exercício 26
Indexação de arrays 3-D



Acesse os números 5, 9, 3 e 21 do array abaixo.



- import numpy as np
- ar = np.array([
-     [[1, 1, 5, 1],
-      [1, 1, 1, 9]],
-     [[1, 3, 1, 1],
-      [1, 1, 21, 1]]
              ])

In [191]:
import numpy as np

ar = np.array([
    [[1, 1, 5, 1],
    [1, 1, 1, 9]],
    [[1, 3, 1, 1],
    [1, 1, 21, 1]]
])

ar

array([[[ 1,  1,  5,  1],
        [ 1,  1,  1,  9]],

       [[ 1,  3,  1,  1],
        [ 1,  1, 21,  1]]])

In [197]:
print(ar[0][0][2])
print(ar[0][1][-1])
print(ar[1][0][1])
print(ar[1][1][2])

5
9
3
21


# Exercício 27
Slicing 1-D

Com base no array abaixo obtenha o slice do índice 2 até o índice 6.



- import numpy as np
- ar = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])


Dica: Em Python, slicing significa que podemos selecionar elementos entre índices especificados. Sintaxe: [start:end:pass].

Se não especificarmos o start o valor default será 0. Se não definirmos o step, será considerado o valor 1.

In [198]:
import numpy as np
ar = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
ar

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [201]:
print(ar[1:10])

[ 2  3  4  5  6  7  8  9 10]
